Submit a plugin for scanning container/image

carbonetes opened the pull request:
<!--

PLUGIN DEVELOPERS: If you are submitting a new plugin

- Make sure you read the Plugin Naming Guide: https://krew.sigs.k8s.io/docs/developer-guide/develop/naming-guide/
- Verify you can install your plugin locally: kubectl krew install --manifest=[...] --archive=[...]

-->

carbonetes commented:
I've signed it. Thanks.

carbonetes commented:
/assign @chriskim06
/assign @corneliusweig

ahmetb commented on plugins/carbonetes-scan.yaml:
```diff
@@ -0,0 +1,30 @@
+  name: carbonetes-scan
+spec:
+  homepage: https://github.com/carbonetes/kubectl-carbonetes-scan
+  shortDescription: Seamlessly integrates complete container analysis directly into your Kubernetes cluster.
```
shortening by half would be ideal, lots of redundant/subjective words in there.

carbonetes replied:
Thanks for the review. We'll make a change to this description.

ahmetb commented on plugins/carbonetes-scan.yaml:
```diff
@@ -0,0 +1,30 @@
+  homepage: https://github.com/carbonetes/kubectl-carbonetes-scan
+  shortDescription: Seamlessly integrates complete container analysis directly into your Kubernetes cluster.
+  version: v1.0.0
+  description: |
```
by examining your source code I can tell there's a huge caveat in this plugin: This plugin requires Docker on your machine to work.

Make use of `caveats:` section and check if `docker` is installed, otherwise error properly please.

ahmetb replied:
also, what's the reason why you don't distribute the `carbonetes` binary natively for linux/mac and just let users use that? what's the point of docker-run in this case?

carbonetes replied:
The reason why we used docker-run is that we already have an existing docker image on the docker hub, which we were currently using on our plugins on other software development tools. We want to maintain these plugins simultaneously when we have some new features, so we've decided to use the same image for creating other plugins.

ahmetb commented:
So it seems this is a very light shim on top of a `docker run` command. Typically we reject plugins that are just light shims on top of a heavyweight process (such as those installing software, or bootstrapping programs).

kubectl plugins typically do meaningful work standalone, but in this case this plugin is parsing an image name from a pod and passing it to `docker run [carbonetes-image]` in a shell script. I'm not sure if that's enough value add, but let's debate.
/kind gray-area
/cc @corneliusweig

carbonetes commented:
> So it seems this is a very light shim on top of a `docker run` command. Typically we reject plugins that are just light shims on top of a heavyweight process (such as those installing software, or bootstrapping programs).
> 
> kubectl plugins typically do meaningful work standalone, but in this case this plugin is parsing an image name from a pod and passing it to `docker run [carbonetes-image]` in a shell script. I'm not sure if that's enough value add, but let's debate.
> /kind gray-area
> /cc @corneliusweig

I've already updated the manifest file and make use of the **caveats:** section. Also, we added checking if Docker is installed as you prescribed. I hope these changes add some weight for your approval.

corneliusweig commented:
Thanks for clarifying this and making the edits.

However, I'd still like to understand why this plugin needs to be run via `docker run`. The much preferred way would be to ship a monolithic binary that contains the logic.

One drawback of your current approach is that you don't know which version is actually being executed. You run `docker run carbonetes/carbonetes-scan:latest`. This is bad for two reasons:

- when running for the first time, you don't know which version will be pulled
- if the image was pulled before, it won't be pulled again

The bottom line is that you never know what version really runs when a user has problems with our plugin.

---

If you insist that distribution via docker containers is the only viable approach for you, I recommend to:

- specify the exact version tag in your script, ideally with the version sha:  `docker run yourimage@sha256:12345fe`
- or use the `--pull=always` option of docker >=19.09 (not so nice because it downloads the image every time!)

carbonetes commented:
> Thanks for clarifying this and making the edits.
> 
> However, I'd still like to understand why this plugin needs to be run via `docker run`. The much preferred way would be to ship a monolithic binary that contains the logic.
> 
> One drawback of your current approach is that you don't know which version is actually being executed. You run `docker run carbonetes/carbonetes-scan:latest`. This is bad for two reasons:
> 
> * when running for the first time, you don't know which version will be pulled
> * if the image was pulled before, it won't be pulled again
> 
> The bottom line is that you never know what version really runs when a user has problems with our plugin.
> 
> If you insist that distribution via docker containers is the only viable approach for you, I recommend to:
> 
> * specify the exact version tag in your script, ideally with the version sha:  `docker run yourimage@sha256:12345fe`
> * or use the `--pull=always` option of docker >=19.09 (not so nice because it downloads the image every time!)

Hi @corneliusweig,

Since you want kubectl plugins to do significant work independently, we have to make changes to our code. We'll remove the **docker run** if that makes our plugin incorrect.

Our apologies for insisting to get approval although in the first place you've already mentioned that there's something wrong with our code.

The next thing is what we'll be going to do with this PR? Do we just close this or you need to reject this PR?

Or we'll just need to update our code for this PR?

ahmetb commented:
I think the recommended action would be to distribute your plugin directly as a binary (like other plugins do) as opposed to a script that eventually runs "Docker run"￼.


If that doesn't work for you, we can close this PR, but can still distribute your plugin via Krew using custom indexes: https://krew.sigs.k8s.io/docs/developer-guide/custom-indexes/ That way users can still install your plugin via Krew, but they'll be downloading the plugin manifest from your repository (instead of krew-index).

carbonetes commented:
Hi, @ahmetb and @corneliusweig,

We've now removed the **docker run** on our code as you prescribed.

Please review again our code.

Thanks.

corneliusweig commented on plugins/carbonetes-scan.yaml:
```diff
@@ -5,15 +5,17 @@ metadata:
     license types, malware, secrets, and bill of materials.
   caveats: |
-    Docker needs to be installed on your machine to use this plugin.
+    You need to have a valid credentials at Carbonetes to execute this plugin.
+    Doesn't have any credentials yet?
+    Register here: https://carbonetes.com
```
🤔 This seems to be a new requirement. Is that really necessary?
I think no existing plugin requires an account for a third-party site. I was looking if we explicitly require that plugins are open source, but could not find it. Let's hear more opinions.
/cc @ahmetb @chriskim06

corneliusweig replied:
Ok, I think an account as prerequisite is fine. However, please make it  clear that the evaluated manifests are sent to a third party.

corneliusweig reviewed (commented):
Do I understand correctly, that the analysis works by sending a bunch of details to an analysis server?
https://github.com/carbonetes/kubectl-carbonetes-scan/blob/66d21034c9a305c5a2dcefc396bbb8a1cd9cfd19/kubectl-carbonetes-scan#L485

This needs to be mentioned in `caveats` section, because the default assumption is that the only communication is between the k8s control plane and the plugin. But not third parties.

ahmetb commented:
@corneliusweig I think quite a few plugins by AquaSecurity or ReplicatedHQ follow the same model (sending your data to a remote server or requiring an account).

corneliusweig commented:
> @corneliusweig I think quite a few plugins by AquaSecurity or ReplicatedHQ follow the same model (sending your data to a remote server or requiring an account).

IIUC that's not the case. The closest I could find was in `support-bundle`:

> This archive can be manually inspected or uploaded to https://vendor.replicated.com for automated analysis.

So no automatic upload, and still a choice for manual inspection. There's also aquasecurity/starboard, which builds on trivy---but AFAICT this operates locally.

---

@carbonetes In any case, before we can continue evaluating your submission, you have to sign the Kubernetes CLA. 😉

carbonetes commented:
I signed it

carbonetes commented:
Hi,

Any updates on this?

Thanks.

corneliusweig commented on plugins/carbonetes-scan.yaml:
```diff
@@ -0,0 +1,34 @@
+  version: v1.1.0
+  description: |
+    Provides complete container analysis and
+    policy evaluation as a fully managed service.
```
What kind of policies is that? Do users have to specify them?

corneliusweig commented on plugins/carbonetes-scan.yaml:
```diff
@@ -0,0 +1,34 @@
+  homepage: https://github.com/carbonetes/kubectl-carbonetes-scan
+  shortDescription: Integrates container analysis directly into your cluster.
+  version: v1.1.0
+  description: |
```
Please wrap around at ~80 chars, so that it prints nicely on wide and narrow terminals.

corneliusweig reviewed (commented):
Sorry, this slipped through the cracks. 🐌 

I think we need to be open for commercial plugins, or  plugins that require accounts with a third party. However, please state clearly that your plugin sends manifests to a third party service.

/cc @ahmetb

ahmetb commented:
> However, please state clearly that your plugin sends manifests to a third party service.

We have several replicatedhq plugins also work in a similar fashion. I am not sure if this warrants an explicit statement, but it's always good to have.
